### ISSUE TITLE
Enable optional name change and external identity to be surfaced for auth'ed chat accounts

### ISSUE THREAD

**Issue by brokenintuition** [owncast/owncast#2354]
### Share your bug report, feature request, or comment.

I was watching a stream (still am actually) where most of the viewers in chat changed their name to match their full @username@instance fediverse name. I think it would be good to make that an option to change automatically when using fediauth.

It might also be good for name changes where the new name looks like a fediverse account name for owncast to enforce that the user is authenticated as that account to avoid confusion with the check mark looking like validation that the person in the chat is that owner of that account.

I haven't thought about implementation yet, just wanted to open the issue for discussion before I forgot about it :)

**Comment by gabek** [comment]
This is a good discussion, and things I've thought about in the past.

Originally I wanted to directly tie somebody's chat account to their Fediverse identity when they authenticate through the Fediverse with a link to their profile, image, and everything. But the more I thought about the more I realized that it wasn't a fair decision to essentially "dox" somebody's identity in one place just because they used it to authenticate themselves in another.

And while I could certainly add more buttons and checkboxes and pieces of text to allow people to opt-in, the whole feature is already quite complex and confuses people, and this would only make it worse. So while I wish I could do all the cool bells and whistles, I think I just have to leave it to the individual to share what they want to share and have their name be what they want it to be without me making those choices for them. Though I think this would be very cool and I wish I could add it.

As for stopping people from changing their name to something that looks like a Fediverse account, while that sounds cool, it would almost be impossible to stop it. It's the "looks like" part that is tricky. You can do a lot of clever things with text to make something look like another.

**Comment by aaronpk** [comment]
I actually had this same thought yesterday too when logging in to a stream with my IndieAuth URL. I was actually surprised that it *didn't* change my name in the chat. At first I actually thought the login failed, since there was almost no indication that anything had changed on my end.

What would be the reason someone would authenticate if they didn't want to share their name? Seems easier to just not authenticate.

**Comment by gabek** [comment]
I just have a hard time making the decision that just because somebody is using an authentication method it means they want to publicly share an identity. 

> What would be the reason someone would authenticate if they didn't want to share their name? Seems easier to just not authenticate.

While not a 1:1 example, it's like if you use your email or phone number to sign up for something, it doesn't share that with everyone else on the service. Or if you "Login with Facebook" it doesn't give everyone your Facebook account. (Though I think services used to be far more liberal with that in the past).

I understand I'm being a bit conservative in this view, and adding this kind of functionality would be great for 99.9% of people, and it's what they'd want. But I fear for that one person who did not intend to share their identity to an entire chat and the fallout that could cause for the person and the project.

I do think opt-ins, more text, etc could solve this problem. But admittedly I don't feel comfortable coming up with a solution that 1) Actually solves the problem and 2) Solves it without making the entire flow and concept more confusing than it already is. UX isn't my strength, and doing it wrong could be a big problem.

**Comment by aaronpk** [comment]
I think if anonymous chat wasn't an option I would agree with this. But the fact that you can just chat and even change your username without any login means someone can just stay private that way.

There's also a big difference between a Fediverse/Web URL than email or phone number. Someone's Fediverse/website is generally used as a publicly facing thing, or their account is protected and only shows things to followers. In both cases they are *intended* on being used as a public identifier though. "Find me on the Fediverse, here's my address" where the person could then choose to accept a follow request. Email and phone are different, those are usually treated as a private communication medium, and only given out after someone agrees to be contacted that way.

I think your facebook example is wrong though, think of comment services like Disqus, if you log in with facebook to comment then it will link to your profile that you used to log in and show your name from that service.

I also agree more checkboxes and options are likely to be confusing.

I don't know how to survey people about this, but it feels totally fine having the authentication also trigger identifying the user in the chat, since you can just not authenticate if you don't want to share your identity.

**Comment by aaronpk** [comment]
One thought that just occurred to me. Since I really did struggle to understand whether my login was successful, what about putting a prompt somewhere (maybe inline in the chat box?) after logging in that says "Login successful, do you want to change your chat username to **X**?"

**Comment by brokenintuition** [comment]
> I do think opt-ins

Apologies if it wasn't clear in my original description, I was intending it to be an opt-in behavior. I was thinking a check box on the auth page/success dialog but I agree it's not easy to add things without being confusing. UX is not my strength either.

> It's the "looks like" part that is tricky.

Yeah I realized right after posting email is a pretty good counterexample to what I was thinking. Without the leading "@" there's no way to tell if user@domain is a fediverse address or an email address and (although I can't imagine why anybody would change their name to an email address) it'd be pretty confusing if your name change was denied for being a fediverse account name if you have no idea what that is.

**Comment by gabek** [comment]
I guess it's also partly a mindset thing on my side. Chat auth was never about global identity, it was about having a way to restore your *Owncast* chat identity, in order to keep your moderator status, or whatever goal you have in keeping the same chat identity between devices, browsers, etc. That was the core problem that was being solved.

And as hesitant as I am, I do think people should marginally expect that if you share Identity A with Service B that it will be used for something, like in Aaron's Disqus example. And if you don't want _anybody_ to know Identity A then you shouldn't be throwing it around elsewhere. I don't actually know if anybody could disagree with that.

I think the discussion about post-auth options is a good one. It allows the auth process itself to be no more complex than it already is, but simply opens up a couple new options afterwards.

Expanding on @aaronpk's thoughts. In the new version of the Owncast UI the name change flow is actually in a modal where it also allows you to change your display color. How would we feel about after auth is complete you get provided this modal with an updated name pre-populated, and a new option presented that's like "Share external link" or something. It can be determined if it should be enabled by default. But in this modal the person then has the option to either close the modal and do nothing, or to save the new name that was populated based on the auth data.

New modal (obviously not yet complete): https://owncast.online/components/?path=%2Fstory%2Fowncast-modals-name-change--basic


<blockquote><div><strong><a href="https://owncast.online/components/?path=%2Fstory%2Fowncast-modals-name-change--basic">Webpack App</a></strong></div></blockquote>

**Comment by gabek** [comment]
I'd have to remove the:

> Note: This is for authentication purposes only, and no personal information will be accessed or stored.

In auth modals!

**Comment by brokenintuition** [comment]
> How would we feel about after auth is complete you get provided this modal with an updated name pre-populated

Sounds great to me. I like separating the "name from auth" idea from the auth workflow itself. I don't have any strong opinion on whether it should be enabled by default or not.

**Comment by gabek** [comment]
I added a "needs design" tag so hopefully somebody can work out what this should look like since it can get more complex than it was probably initially thought. Shoving a bunch of repeated logos next to somebody's name probably isn't the solution. As always, anybody with a design skill set would be very appreciated!

- [ ] What does it look like with two sources? For example one IndieAuth and one FediAuth.
- [ ] What does it look like with two sources of the same source? For example two FediAuth.
- [ ] What does it look like with many sources? For example 5 Fediverse accounts.

**Comment by valvin1** [comment]
Hi, I'm joining this discussion after having a viewer not understanding what brings fediverse authentication. I'm not really aware of IndieAuth so I won't discuss directly about it.

To be honest, I believed it was a sort of login to indicates I'm the true `username@domain.tld`. By reading this discussion I understand it is absolutely not what it was designed.

So I'm wondering what are the use case of authentication as a viewer, moderator or streamer?

It looks there is no existing settings granting specific rights after authentication?

* As a streamer I can grant moderator to an authenticated user?
* As a streamer or moderator (I've not well used this feature) if a user is authenticated I can check if the message has been sent by a specific user?
* As a viewer it brings an indicator (shield on nightly) that I've been authenticated. Streamer (and maybe moderator) can see who am I.

It is possible I'm wrong on the feature behind the authentication do not hesitate to correct me.

I'll stream tonight to understand better the features.

**Comment by gabek** [comment]
The use case for this feature has always been Owncast chat authentication. Here is an example:

1. Bob joins Owncast chat from his computer and it gives him the name "random-person".
2. Bob changes name to "cool-bob".
3. Bob is given moderator rights.
4. Bob authenticates so he can be the same chat user from elsewhere.
5. Bob joins Owncast chat from his phone and it gives him the name "different-name".
6. Bob authenticates from his phone.
7. Bob is now "cool-bob" and is a moderator on his phone and his computer.

**Comment by valvin1** [comment]
I think I was totally wrong on what is the purpose of the authentication.

I did some test and see that streamer and moderator doesn't have any indication of authentication information.

So the main goal of the authentication is to register a nickname. So the first use case could be for the streamer and register its own nickname and grants moderator rights to this nickname. Thanks to the authentication no one else could take this nickname until it authenticates using a/the known authentification provider.

In user details I see several nickname for my own user. Does it means all nicknames I have used are now registered or only the last one?

**Comment by gabek** [comment]
It's not for registering a nickname, it's for restoring your identity. Locking down your current nickname (you can only have one at a time) was added because people asked for it.

It's primarily for people who are moderators so they can continue to be a moderator from other devices. But it also shows people that the "cool-bob" in the chat is the same "cool-bob" they saw previously.

**Comment by gabek** [comment]
Maybe I can clarify. The word "identity" here may be confusing as it's easy to think "identity" means "nickname". But "identity" is all of the characteristics and metadata that make up an Owncast chat account. Display name is one of them, color is one of them, date created is one of them, moderator state is one of them, etc etc.

This is what you want to be able to restore when moving to a different browser or different device.

**Comment by valvin1** [comment]
So authentication allowes user to save its profile / chat settings on the owncast server. As a moderator it should be highly recommended.

I think the first thing I see as an user is "Authenticate with chat". This wording let me think (wrongly) bring your profile from the Fediverse and it looks I'm not the only one. Maybe rewording this by something like "Register / Bring back your chat profile" could help ? By the way it looks not so simple.

I think I'm a bit out of the topic of this issue as this one deal about bringing nickname from the external identity server.

About this topic, maybe the modal mentioned could be only popup when we are using a generated nickname (don't know if it is known by owncast)?
